src/coupled_driver.cpp

```diff
@@ -267,216 +290,274 @@ bool CoupledDriver::is_converged()
   double norm;
 
   // The heat root has the global temperature data
+  norm = this->temperature_norm(norm_);
   if (comm_.rank == heat_root_) {
-    norm = this->temperature_norm(norm_);
     converged = norm < epsilon_;
   }
 
   comm_.broadcast(converged, heat_root_);
   comm_.broadcast(norm, heat_root_);
 
-  std::string msg = "temperature norm: " + std::to_string(norm);
-  comm_.message(msg);
+  std::stringstream msg;
+  msg << "temperature norm: " << norm;
+  comm_.message(msg.str());
   return converged;
 }
 
 void CoupledDriver::update_heat_source(bool relax)
 {
+  comm_.message("Updating heat source");
   auto& neutronics = this->get_neutronics_driver();
   auto& heat = this->get_heat_driver();
 
-  // ****************************************************************************
-  // Gather heat source on neutronics root and apply underrelaxation
-  // ****************************************************************************
-
-  if (relax && comm_.rank == neutronics_root_) {
-    // Store previous heat source solution if more than one iteration has been performed
-    // (otherwise there is not an initial condition for the heat source)
-    std::copy(heat_source_.begin(), heat_source_.end(), heat_source_prev_.begin());
+  if (relax && heat.active()) {
+    std::copy(cell_heat_.cbegin(), cell_heat_.cend(), cell_heat_prev_.begin());
   }
 
+  decltype(cells_) cells_recv;
+  decltype(cell_heat_) cell_heat_send;
+  xt::xtensor<double, 1> all_cell_heat;
+
+  // For the coupling scheme, only the neutronics root needs the heat source.
+  // However, to compute the heat source, OpenmcDriver::heat_source must
+  // do a collective operation on all the ranks in the neutronics sub comm.
+  // Hence, all neutronics ranks must call OpenmcDriver::heat_source
```

Comment:
The coupling itself only requires that we have the global cell heat sources on the neutronics root.  This is because we send/recv from the neutronics root to each heat rank.  However, the `OpenMC` driver needs to do a collective operation to return the global cell heat sources, so we still need to call `neutronics.heat_source` on all neutronics procs.  

I think there's an opportunity to do overlap this communication a bit.  That is, we could probably partition the heat ranks between the neutronics ranks; and then do multiple send/recvs simultaneously.  I might try that in another PR; just wanted to get this correct before I started getting fancy.